[FLINK-28420] Support partial caching in sync and async lookup runner

PatrickRen opened the pull request:
## What is the purpose of the change

This pull request add partial caching functionality in sync and async lookup runners, which is a part of FLIP-221.

## Brief change log
- Introduce CachingLookupFunction and CachingAsyncLookupFunction as wrapper of lookup function with caching layer
- Port TestValuesLookupFunction onto new LookupFunction API and add tests for sync lookup
- Port AsyncTestValuesLookupFunction onto new AsyncLookupFunction and add tests for async lookup

## Verifying this change

This change modifies some existing tests, including *LookupJoinHarnessTest* and *LookupJoinITCase*.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (**yes** / no)
  - If yes, how is the feature documented? (not applicable / docs / **JavaDocs** / not documented)

wuchong commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/table/lookup/CachingLookupFunction.java:
```diff
@@ -0,0 +1,147 @@
+ * after lookup for later use.
+ */
+@Internal
+public class CachingLookupFunction extends LookupFunction {
```
Declare `serialVersionUID` field.

wuchong commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/table/lookup/CachingAsyncLookupFunction.java:
```diff
@@ -0,0 +1,95 @@
+import java.util.Collections;
+import java.util.concurrent.CompletableFuture;
+
+public class CachingAsyncLookupFunction extends AsyncLookupFunction {
```
Declare `serialVersionUID` field.

wuchong commented on flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/runtime/stream/sql/AsyncLookupJoinITCase.scala:
```diff
@@ -305,19 +320,147 @@ class AsyncLookupJoinITCase(
+        ROCKSDB_BACKEND,
+        ENABLE_OBJECT_REUSE,
+        AsyncOutputMode.ALLOW_UNORDERED,
+        ENABLE_CACHE)
```
I think we don't need double cases for dynamic table sources. I think mixing cache cases into the existing 4 cases (for dynamic table sources) is enough.

PatrickRen replied:
I'll remove the ROCKSDB_BACKEND cases with caching as these two seems unrelated.

PatrickRen commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/table/lookup/CachingAsyncLookupFunction.java:
```diff
@@ -0,0 +1,95 @@
+                    .whenComplete(
+                            (lookupValues, throwable) -> {
+                                if (throwable != null) {
+                                    numLoadFailuresCounter.inc();
```
I'll leave it as is and add a TODO here because FLIP-234 will handle the retrying logic.

PatrickRen commented:
Really appreciated for the review during the weekend @wuchong ! I addressed your comments in b78bb13726e5a568c08dcd06c8bfb8293a7a50a9. Please take another look when you are available~

wuchong reviewed (approved):
LGTM.

PatrickRen commented:
@flinkbot run azure

wuchong commented:
The failed case is tracked by FLINK-28844. 
Merging...
